From bc8b91ffc5b746250190b8f4244373ab4cdcebae Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Tue, 28 Jul 2026 23:33:44 +0800
Subject: [PATCH] mtk: mt76: mt7996: disable beacons when going offchannel

Porting from wifi: mt76: mt7996: disable beacons when going offchannel

When AP and STA interfaces are operating concurrently, the STA interface
may perform a scan or remain-on-channel operation.
In these cases, mac80211 updates the RF channel in the firmware for the
STA off-channel activity.
However, the firmware beacon queue continues to transmit beacons without
validating the current operating channel.
As a result, AP beacons can be sent on the wrong channel while the STA is
off-channel, potentially causing interoperability issues and violating
regulatory requirements.

This patch disables beacon transmission when going off-channel, ensuring
that AP beacons are only sent on the correct operating channel.

Additionally, add the offchan_stop_bcn flag to avoid setting the beacon during
a normal CSA.

Without this flag, the beacon will be updated after every RF channel switch.
During a normal CSA flow, because the beacon template is not updated to
the csa_after beacon and offs->cntdwn_counter_offs is not yet reset, another
countdown will be restarted, which affects the subsequent channel switch.

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 mt76.h        | 1 +
 mt7996/main.c | 9 +++++++++
 mt7996/mcu.c  | 3 +++
 3 files changed, 13 insertions(+)

diff --git a/mt76.h b/mt76.h
index 7af33cc04..6a629d23b 100644
--- a/mt76.h
+++ b/mt76.h
@@ -1120,6 +1120,7 @@ struct mt76_phy {
 	struct cfg80211_chan_def chandef;
 	struct cfg80211_chan_def main_chandef;
 	bool offchannel;
+	bool offchan_bcn_stopped;
 	bool radar_enabled;
 
 	struct delayed_work roc_work;
diff --git a/mt7996/main.c b/mt7996/main.c
index 8c0cbc541..01e55d877 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -820,6 +820,11 @@ int mt7996_set_channel(struct mt76_phy *mphy)
 			goto out;
 	}
 
+	if (mphy->offchannel) {
+		mt7996_mac_update_beacons(phy);
+		mphy->offchan_bcn_stopped = true;
+	}
+
 	ret = mt7996_mcu_set_chan_info(phy, UNI_CHANNEL_SWITCH, false);
 	if (ret)
 		goto out;
@@ -840,6 +845,10 @@ int mt7996_set_channel(struct mt76_phy *mphy)
 		mt7996_mcu_edcca_threshold_ctrl(phy, phy->user_edcca_threshold, true);
 #endif
 	phy->noise = 0;
+	if (!mphy->offchannel && mphy->offchan_bcn_stopped) {
+		mt7996_mac_update_beacons(phy);
+		mphy->offchan_bcn_stopped = false;
+	}
 
 out:
 	ieee80211_queue_delayed_work(mphy->hw, &mphy->mac_work,
diff --git a/mt7996/mcu.c b/mt7996/mcu.c
index 49ea3e31d..6c6cce55f 100644
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -5098,6 +5098,9 @@ int mt7996_mcu_add_beacon(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 	if (!mlink)
 		return -EINVAL;
 
+	if (link->phy && link->phy->mt76->offchannel)
+		enabled = false;
+
 	rskb = __mt7996_mcu_alloc_bss_req(&dev->mt76, mlink,
 					  MT7996_MAX_BSS_OFFLOAD_SIZE);
 	if (IS_ERR(rskb))
-- 
2.45.2

